cpp/src/decisiontree/batched-levelalgo/metrics.cuh

```diff
@@ -106,18 +106,16 @@ class GiniObjectiveFunction {
     }
     return sp;
   }
-  static DI LabelT LeafPrediction(BinT* shist, int nclasses)
+  static DI void SetLeafVector(BinT* shist, int nclasses, DataT* out)
   {
-    int class_idx = 0;
-    int count     = 0;
+    // Output probability
+    int total = 0;
     for (int i = 0; i < nclasses; i++) {
-      auto current_count = shist[i].x;
-      if (current_count > count) {
-        class_idx = i;
-        count     = current_count;
-      }
+      total += shist[i].x;
+    }
+    for (int i = 0; i < nclasses; i++) {
+      out[i] = DataT(shist[i].x) / total;
```

Comment:
This division can be avoided by first computing the inverse of `total`. Not sure if compiler already figures it out. I haven't looked at SASS.
Same for other two `SetLeafVector` functions.

Reply:
This calculation is performed only once per leaf, per class. Are you sure this optimisation is necessary?